src/saveload/saveload.cpp

```diff
@@ -1217,141 +1218,144 @@ static void *IntToReference(size_t index, SLRefType rt)
+	 * @param storage The storage to find the size of
 	 * @param conv VarType type of variable that is used for calculating the size
+	 * @param cmd The SaveLoadType ware are saving/loading.
 	 */
-	static size_t SlCalcDequeLen(const void *deque, VarType conv)
+	static size_t SlCalcLen(const void *storage, VarType conv, SaveLoadType cmd = SL_VAR)
 	{
-		const SlDequeT *l = (const SlDequeT *)deque;
+		assert(cmd == SL_VAR || cmd == SL_REF);
 
-		int type_size = 4;
-		/* Each entry is saved as type_size bytes, plus type_size bytes are used for the length
-		 * of the list */
-		return l->size() * SlCalcConvFileLen(conv) + type_size;
+		const SlStorageT *list = static_cast<const SlStorageT *>(storage);
+
+		int type_size = 4; // Size of the length of the list.
+		int item_size = cmd == SL_VAR ? SlCalcConvFileLen(conv) : (IsSavegameVersionBefore(SLV_69) ? 2 : 4);
+		return list->size() * item_size + type_size;
+	}
+
+	static void SlSaveLoadMember(SaveLoadType cmd, Tvar *item, VarType conv)
+	{
+		switch (cmd) {
+			case SL_VAR: SlSaveLoadConv(item, conv); break;
+			case SL_REF: SlSaveLoadRef(item, conv); break;
+			default:
+				NOT_REACHED();
+		}
 	}
 
 	/**
-	 * Internal templated helper to save/load a std::deque.
-	 * @param deque The std::deque being manipulated
-	 * @param conv VarType type of variable that is used for calculating the size
-	*/
-	static void SlDeque(void *deque, VarType conv)
+	 * Internal templated helper to save/load a list-like type.
+	 * @param storage The storage being manipulated.
+	 * @param conv VarType type of variable that is used for calculating the size.
+	 * @param cmd The SaveLoadType ware are saving/loading.
+	 */
+	static void SlSaveLoad(void *storage, VarType conv, SaveLoadType cmd = SL_VAR)
 	{
-		SlDequeT *l = (SlDequeT *)deque;
+		assert(cmd == SL_VAR || cmd == SL_REF);
+
+		SlStorageT *list = static_cast<SlStorageT *>(storage);
 
 		switch (_sl.action) {
-			case SLA_SAVE: {
-				SlWriteUint32((uint32)l->size());
+			case SLA_SAVE:
+				SlWriteUint32((uint32)list->size());
 
-				typename SlDequeT::iterator iter;
-				for (iter = l->begin(); iter != l->end(); ++iter) {
-					SlSaveLoadConv(&(*iter), conv);
+				for (auto &item : *list) {
+					SlSaveLoadMember(cmd, &item, conv);
 				}
 				break;
-			}
+
 			case SLA_LOAD_CHECK:
 			case SLA_LOAD: {
-				size_t length = SlReadUint32();
+				size_t length;
+				switch (cmd) {
+					case SL_VAR: length = SlReadUint32(); break;
+					case SL_REF: length = IsSavegameVersionBefore(SLV_69) ? SlReadUint16() : SlReadUint32(); break;
+					default: NOT_REACHED();
+				}
 
-				/* Load each value and push to the end of the deque */
+				/* Load each value and push to the end of the storage. */
 				for (size_t i = 0; i < length; i++) {
-					T data;
-					SlSaveLoadConv(&data, conv);
-					l->push_back(data);
+					Tvar data;
+					SlSaveLoadMember(cmd, &data, conv);
+					list->push_back(data);
```

Comment:
Could be something like `Tvar &data = list->emplace_back();` and then the `list->push_back(data);` is not needed anymore. Less code and less copying, as it will immediately write the data in the element in the storage instead of first on the stack and then copying it to the storage.